### ISSUE TITLE
Tern crash on file deletion

### ISSUE THREAD

**Issue by vincentwoo** [ternjs/tern#764]
Hi, this is going to be a bit of a complicated bug report.

# TLDR

On sending a request to delete a file from Tern's cache, I can cause the following Tern crash to occur:

```
/home/coderpad/node_modules/tern/lib/tern.js:502
      parent = srv.fileMap[parent].parent;
                                  ^

TypeError: Cannot read property 'parent' of undefined
    at parentDepth (/home/coderpad/node_modules/tern/lib/tern.js:502:35)
    at /home/coderpad/node_modules/tern/lib/tern.js:372:16
    at Array.sort (native)
    at analyzeAll (/home/coderpad/node_modules/tern/lib/tern.js:371:17)
    at /home/coderpad/node_modules/tern/lib/tern.js:165:11
    at /home/coderpad/node_modules/tern/lib/tern.js:272:9
    at Object.exports.withContext (/home/coderpad/node_modules/tern/lib/infer.js:872:18)
    at /home/coderpad/node_modules/tern/lib/tern.js:263:13
    at analyzeAll (/home/coderpad/node_modules/tern/lib/tern.js:392:15)
    at doRequest (/home/coderpad/node_modules/tern/lib/tern.js:256:5)
```

# Background

I use Tern as a remote server for autocompleting arbitrary blobs of JS code which are not tied to the local filesystem in any way. I've been trying to track problems in production I've been seeing with Tern consuming 100% CPU and a ton of memory. My working hypothesis is that this is due to the way I currently use Tern: assign each request a sequential ID, and tell Tern about a new file each time with that ID. 

Loosely speaking, that strategy looked like this:

```coffee
reqID = 0
callTern = ({pad, line, ch, code}, cb) ->
  filepath = "/home/coderpad/#{++reqID}.js"

  doc = {
    query:
      type: 'completions'
      file: '#0'
      end: { line, ch }
      types: true
      caseInsensitive: true
      lineCharPositions: true
    files: [
      type: 'full'
      name: filepath
      text: code
    ]
    timeout: 5000
  }

  request.post 'local tern server', json: doc, (error, response, results) ->
    cb(results)
```

As you can see this would add a new filename (like `/home/coderpad/14.js`) each time, which I am given to understand Tern does not automatically clean out. Therefore, I added the following bit of code to try to see the effect on performance:

```coffee
# ... same as above
request.post 'local tern server', json: doc, (error, response, results) ->
  deletion = {
    files: [
      type: 'delete'
      name: filepath
    ]
  }
  request.post 'local tern server', json: deletion, (error, response, results) ->
    # ...
```

I than reran my integration test of data I've collected from production. I have 400 requests that I distilled because they were able to reliably cause the slowdown from before, and in this case the crash always seems to occur at a specific test case.

# Problems

That it occurs at a specific case was encouraging, but I found that deleting other test cases or changing the order of runs seemed to cause the problem to not happen. I also was unable to cause the problem on OSX running various versions of Node. I could only get this problem to happen in a Docker container on Ubuntu, which is very confusing. The problem seems to require the `node` plugin for tern to be enabled, and I suspect part of the issue has something to do with module resolution.

@marijnh, do you have any ideas about what could be happening? I'd also be happy to send you the test setup I have so far via email, but I can't guarantee it will reproduce the problem on your machine, unfortunately.

**Comment by vincentwoo** [comment]
Okay, I figured out what was making the repro case nonreproducible. It was the filepath set in the query: 

```coffee
filepath = "/home/coderpad/#{++reqID}.js"
...
    files: [
      type: 'full'
      name: filepath
      text: code
    ]
```
Tern uses the filepath set to do the resolving. To repro, I have Tern running in the same directory as the relevant `node_modules`, as well as having the filepath set to that same directory. For example, I'd run `tern` in `/home/coderpad`, ensure that the modules were installed at `/home/coderpad/node_modules`, and instruct Tern via API to add files at, say, `/home/coderpad/14.js`.

If you write files to something like `/tmp/14.js`, everything works okay except for the node modules are not completed.

**Comment by marijnh** [comment]
Attached patch should help with the crash. Could you test it?

**Comment by vincentwoo** [comment]
I think you made a typo:

```
/Users/vwoo/ubuntu/autocomplete/node_modules/tern/lib/tern.js:142
          else if (this.files[i].parent == name) files[i].parent = null;
                                                 ^

ReferenceError: files is not defined
    at Object.signal.mixin.delFile (/Users/vwoo/ubuntu/autocomplete/node_modules/tern/lib/tern.js:142:50)
    at doRequest (/Users/vwoo/ubuntu/autocomplete/node_modules/tern/lib/tern.js:242:13)
    at Object.signal.mixin.request (/Users/vwoo/ubuntu/autocomplete/node_modules/tern/lib/tern.js:164:7)
    at respond (/Users/vwoo/ubuntu/autocomplete/node_modules/tern/bin/tern:259:10)
    at IncomingMessage.<anonymous> (/Users/vwoo/ubuntu/autocomplete/node_modules/tern/bin/tern:165:32)
    at emitNone (events.js:80:13)
    at IncomingMessage.emit (events.js:179:7)
    at endReadableNT (_stream_readable.js:913:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

However, changing `files[i].parent = null;` to `this.files[i].parent = null;` seems to have fixed the crash issue!

**Comment by marijnh** [comment]
Woops. Thanks for catching that.
